lib/rubocop/cop/rspec/described_class.rb

```diff
@@ -156,64 +156,83 @@ def offensive_described_class?(node)
-        #   const_name(s(:const, s(:const, nil, :M), :C)) # => [:M, :C]
-        #   const_name(s(:const, s(:cbase), :C))          # => [nil, :C]
-        def const_name(node)
-          # rubocop:disable InternalAffairs/NodeDestructuring
-          namespace, name = *node
-          # rubocop:enable InternalAffairs/NodeDestructuring
-          if !namespace
-            [name]
-          elsif namespace.const_type?
-            [*const_name(namespace), name]
-          elsif namespace.lvar_type? || namespace.cbase_type?
-            [nil, name]
+          # @param namespace [Array<Symbol>]
+          # @param const [Array<Symbol>]
+          # @return [Array<Symbol>]
+          # @example
+          #   # nil represents base constant
+          #   collapse_namespace([], [:C])                # => [:C]
+          #   collapse_namespace([:A, :B], [:C])          # => [:A, :B, :C]
+          #   collapse_namespace([:A, :B], [:B, :C])      # => [:A, :B, :C]
+          #   collapse_namespace([:A, :B], [nil, :C])     # => [nil, :C]
+          #   collapse_namespace([:A, :B], [nil, :B, :C]) # => [nil, :B, :C]
+          def collapse_namespace(namespace, const)
+            return const if namespace.empty?
+            return const if const.first.nil?
+
+            start = [0, (namespace.length - const.length)].max
+            max = namespace.length
+            intersection = (start..max).find do |shift|
+              namespace[shift, max - shift] == const[0, max - shift]
+            end
+            [*namespace[0, intersection], *const]
           end
-        end
 
-        # @param node [RuboCop::AST::Node]
-        # @return [Array<Symbol>]
-        # @example
-        #   namespace(node) # => [:A, :B, :C]
-        def namespace(node)
-          node
-            .each_ancestor(:class, :module)
-            .reverse_each
-            .flat_map { |ancestor| ancestor.defined_module_name.split('::') }
-            .map(&:to_sym)
+          # @param node [RuboCop::AST::Node]
+          # @return [Array<Symbol>]
+          # @example
+          #   const_name(s(:const, nil, :C))                # => [:C]
+          #   const_name(s(:const, s(:const, nil, :M), :C)) # => [:M, :C]
+          #   const_name(s(:const, s(:cbase), :C))          # => [nil, :C]
+          def const_name(node)
+            # rubocop:disable InternalAffairs/NodeDestructuring
+            namespace, name = *node
+            # rubocop:enable InternalAffairs/NodeDestructuring
+            if !namespace
+              [name]
+            elsif namespace.const_type?
+              [*const_name(namespace), name]
+            elsif namespace.lvar_type? || namespace.cbase_type?
+              [nil, name]
+            end
+          end
+
+          # @param node [RuboCop::AST::Node]
+          # @return [Array<Symbol>]
+          # @example
+          #   namespace(node) # => [:A, :B, :C]
+          def namespace(node)
```

Comment:
I felt that such a code should belong to rubocop-ast. So I checked and there is indeed `parent_module_name`, which seems to be used only by the `DuplicateMethods` cop and seems to be buggy (or not quite doing the same on purpose). Maybe it would be good to fix it, or at least open a ticket?

Reply:
Filed https://github.com/rubocop-hq/rubocop-rspec/issues/984

Can't agree more that this namespace juggling doesn't really belong here.

@bquorning It seems that with this find the same result of reducing the class length can be achieved by using `parent_module_name`.

Reply:
Well, I actually tried that, with a quick fix in `parent_module_name` itself, but it fails on `::MyClass` - still resolving it under `MyNamespace` instead of the global namespace